### ISSUE TITLE
Problem with `transpose(tan(θ))`

### ISSUE THREAD

**Issue by jfbarthelemy** [sympy/sympy#24205]
Hi all,

I've just found a problem through the use of a library invoking the transpose of matrix blocks. Whenever "blocks" are real, of course the "transpose" should become inoperative. However I have problems with the transpose of the tangent of a symbol which is assumed to be real and which does not simplify.

Here is a minimal script showing the problem (run on Python 3.7.13)

```python
>>> from sympy import symbols, sin, tan, transpose
>>> θ = symbols("θ", real = True)
>>> transpose(tan(θ))
transpose(tan(θ))
>>> transpose(tan(θ)) == tan(θ)
False
>>> transpose(sin(θ))
sin(θ)
>>> transpose(sin(θ)) == sin(θ)
True
```

As you can see the behavior in case of `sin` is the expected one but not with `tan`. Do you have an idea to solve this problem please?

Thanks

**Comment by jksuom** [comment]
`transpose` tries to treat possibly infinite complex arguments as 1 x 1 -matrices. That is tested on this line: https://github.com/sympy/sympy/blob/88b9cba9d7e633ec769baf1fc5267acfd7f54788/sympy/core/expr.py#L1018
Tho problem with `tan(θ)` is that it not always finite even if `θ` is real, so both `is_complex` and `is_infinite` return `None` and the test fails.
```
>>> θ = symbols("θ", real = True)
>>> tan(θ).is_complex
>>> tan(θ).is_infinite
>>> 
```
It seems that there should be an `is_extended_complex` property that would cover both complex numbers and infinity.

**Comment by sbrisard** [comment]
> It seems that there should be an `is_extended_complex` property that would cover both complex numbers and infinity.

How about using the `is_scalar` property? What is discussed here is exactly what defines a scalar, am I wrong ?

Note that the snippet below works as expected

```python
Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 05:59:45) [MSC
v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy
>>> x = sympy.Symbol("x")
>>> y = sympy.tan(x)
>>> y.is_scalar
True
```

**Comment by jksuom** [comment]
It looks like `is_scalar` would work.

**Comment by sbrisard** [comment]
So how do we proceed from here?
Should we submit a PR?

**Comment by jksuom** [comment]
Yes, you could make the change and test it locally. If it works, you can send a PR including tests for `transpose` with arguments like `tan(θ)`, `tan(x)` (with `x = Symbol('x')`),  `cot(x)`, and maybe something else like `1/x`.

**Comment by sbrisard** [comment]
> > It seems that there should be an `is_extended_complex` property that would cover both complex numbers and infinity.
> 
> That's more or less what NumberKind is for.

Agreed. Is `NumberKind` supposed to supersed `is_scalar`? Both seem dangerously similar to me...

**Comment by jfbarthelemy** [comment]
Let me just mention that
```python
>>> from sympy import symbols, sin, tan, transpose
>>> θ = symbols("θ", real = True)
>>> tan(θ).kind
UndefinedKind
>>> tan(θ).is_scalar
True
```

**Comment by oscarbenjamin** [comment]
Here `tan(theta).kind` should be `NumberKind`. It's just that the kind system is still incomplete. I think that there should be a subclass of `Function` called e.g. `NumberFunction` that is defined with NumberKind as domain and codomain. Functions like `tan` should subclass `NumberFunction`.

SymPy has a proliferation of properties like `is_scalar` that don't necessarily mean what the names suggest and whose definitions are often not clearly stated anywhere. In this case a definition is given but it does not seem to bear a clear relationship to `transpose`:
https://github.com/sympy/sympy/blob/88b9cba9d7e633ec769baf1fc5267acfd7f54788/sympy/core/expr.py#L82-L85

The idea of the kind system is that rather than having a many different poorly defined properties that could take contradictory values each object should be associated with a single well defined kind. There should somewhere be a definition of the set of all things that should have that kind so that knowing that an object is of that kind means knowing a well defined fact about the object:
https://github.com/sympy/sympy/blob/88b9cba9d7e633ec769baf1fc5267acfd7f54788/sympy/core/kind.py#L106-L117

**Comment by sbrisard** [comment]
I see. Well, I don't feel confident enough to submit a PR, then. I don't know the `sympy` internals well enough.

**Comment by asmeurer** [comment]
is_scalar is not part of the assumptions system. It's part of the `diff` dispatch system. The assumptions system version of it would be is_finite which means "is finite complex number".

**Comment by sbrisard** [comment]
Which takes us back to @jksuom 's answer, if I am not mistaken.
Is there no clean way to simplify `transpose(tan(x))`?

**Comment by oscarbenjamin** [comment]
I've described a possible approach above. It just requires adding `.kind` to more classes because e.g. `tan` doesn't have it yet. We could also use `is_scalar` but I'm not sure if that muddies the water a bit because the notion of a scalar is not purely about matrices.

More generally though I do wonder why anyone would want to have `transpose(tan(x))` as an expression at all. To me it seems nonsensical to "transpose" something that is not a matrix.

**Comment by sbrisard** [comment]
> More generally though I do wonder why anyone would want to have `transpose(tan(x))` as an expression at all. To me it seems nonsensical to "transpose" something that is not a matrix.

This is a very surprising (and I must admit, a bit offensive) answer. There is nothing nonsensical in calling the transpose of a tensor expression of *any* order (not known before hand). If the tensor order turns out to be 1, it is quite legitimate from a mathematical point of view to expect the transpose to vanish.

**Comment by oscarbenjamin** [comment]
> There is nothing nonsensical in calling the transpose of a tensor expression of _any_ order

The docstring for `transpose` clearly states that its argument should be a matrix and that it represents a matrix:
https://github.com/sympy/sympy/blob/ae895046cd7d5fea6194d46cb51776413ae25f1c/sympy/functions/elementary/complexes.py#L897-L907
The fact that the `transpose` function does not explicitly check that its argument is a matrix is possibly just because without the kind system it is impossible to verify that an object does *not* represent a matrix. Of course `transpose` could be changed if we want to make a more general idea of transposition but that requires clearly defining how everything should be understood. 

The `Expr._eval_transpose` method seems to be based in the idea of linear operators somehow:
https://github.com/sympy/sympy/blob/ae895046cd7d5fea6194d46cb51776413ae25f1c/sympy/core/expr.py#L1016-L1023
The hermitian/antihermitian properties are probably the poorest defined properties in the assumptions system. I'm not aware of anywhere in the codebase that uses them to do anything useful (outside of the matrices module which has its own implementation anyway).

What exactly is the definition of transpose if applied to tensors more generally rather than matrices? Also what do you even mean by "tensors"? Are you referring specifically to the objects that in SymPy are called tensors or to the kind of N-dimensional arrays that appear in numpy/tensorflow etc?

Wikipedia clearly defines the transpose to be an operation on matrices:
https://en.wikipedia.org/wiki/Transpose
That page also says "the transpose of a scalar is the same scalar" but if you look closely that text is just a poor interpretation of the property `transpose(c*A) = c*transpose(A)` which is an equation in which only matrices are transposed. There is no mention on that page of tensors and neither does the page for tensors have any mention of transposition.

**Comment by sbrisard** [comment]
This wikipedia provides a pretty good definition of tensors I believe : https://en.wikipedia.org/wiki/Tensor

A few years ago, I wrote up a small discussion around the transpose of a tensor of the fourth order : https://sbrisard.github.io/posts/20140219-on_the_double_dot_product.html

It might seem far-fetched, but yes, the transpose of a scalar is well-defined: the natural product being seen as a scalar product, then you have

```
x⋅(ay) = (ax)⋅y    for all a ∈ ℝ and x, y ∈ ℝ
```

Note that in the above, `a ∈ ℝ` is seen as the following linear operator over `ℝ` (which is itself seen as a “vector space”): `z ↦ az`; similarly, `x` and `y` are seen as “vectors”.
